src/middlewared/middlewared/plugins/alert.py

```diff
@@ -0,0 +1,444 @@
+from collections import defaultdict
+import copy
+from datetime import datetime
+import json
+import os
+import traceback
+
+from freenasUI.support.utils import get_license
+from licenselib.license import ContractType
+
+from middlewared.alert.base import *
+from middlewared.alert.base import AlertService as _AlertService
+from middlewared.schema import Dict, Str, Bool, Int, accepts, Patch
+from middlewared.service import (
+    ConfigService, CRUDService, Service, ValidationErrors,
+    periodic, private,
+)
+from middlewared.utils import load_modules, load_classes
+
+POLICIES = ["IMMEDIATELY", "HOURLY", "DAILY", "NEVER"]
+DEFAULT_POLICY = "IMMEDIATELY"
+
+alert_sources = []
+alert_services_factories = {}
+
+
+class AlertPolicy:
+    def __init__(self, key=lambda now: now):
+        self.key = key
+
+        self.last_key_value = None
+        self.last_key_value_alerts = defaultdict(dict)
+
+    def receive_alerts(self, now, alerts):
+        gone_alerts = []
+        new_alerts = []
+        key = self.key(now)
+        if key != self.last_key_value:
+            for alert_source_name in set(alerts.keys()) | set(self.last_key_value_alerts.keys()):
+                for gone_alert in (set(self.last_key_value_alerts[alert_source_name].keys()) -
+                                   set(alerts[alert_source_name].keys())):
+                    gone_alerts.append(self.last_key_value_alerts[alert_source_name][gone_alert])
+                for new_alert in (set(alerts[alert_source_name].keys()) -
+                                  set(self.last_key_value_alerts[alert_source_name].keys())):
+                    new_alerts.append(alerts[alert_source_name][new_alert])
+
+            self.last_key_value = key
+            self.last_key_value_alerts = copy.deepcopy(alerts)
+
+        return gone_alerts, new_alerts
+
+
+class AlertService(Service):
+    def __init__(self, middleware):
+        super().__init__(middleware)
+
+        self.node = "A"
+
+        self.alerts = defaultdict(dict)
+
+        self.alert_source_last_run = defaultdict(lambda: datetime.min)
+
+        self.policies = {
+            "IMMEDIATELY": AlertPolicy(),
+            "HOURLY": AlertPolicy(lambda d: (d.date(), d.hour)),
+            "DAILY": AlertPolicy(lambda d: (d.date())),
+            "NEVER": AlertPolicy(lambda d: None),
+        }
+
+    @private
+    async def initialize(self):
+        if not await self.middleware.call("system.is_freenas"):
+            if await self.middleware.call("notifier.failover_node") == "B":
+                self.node = "B"
```

Comment:
I am thinking `node` was a mistake in previous implementation.

Reality is that we probably do not want alert system running in the passive node and we should have truenas alerts running remote middleware calls checking for specific things we want.

Any thoughts on this?

Reply:
Do I understand correctly that it should be like this:

* Passive node does not run alerts by itself

* Active node does remote calls for alert sources marked as e.g. `RUN_ON_PASSIVE_NODE = True`, stores their result in database, sends alerts if needed?

* `node` field itself stays in the database

Reply:
1) Correct

2) Sort of, if you think thats a valid implementation, sure. But here are the current use cases that need check on both nodes: license validation, boot pool status, system version match

3) That or we identify the node with problem in the alert message.

Reply:
`RUN_ON_PASSIVE_NODE = True` would mean "also run on passive node" not "only run on passive node"

Reply:
Yes, my only I fear is if it might complicate things.

How would that work, you run the same code on passive node?

Or would be different alert in the active node that would only run if you're master and do appropriate `failover.call_remote` calls instead?

Reply:
> you run the same code on passive node

Yes. This code can determine node type itself if it is necessary.